### PR TITLE
feat: include notes that have tag titles that match search query

### PR DESCRIPTION
**Pull request by johnny243**
(Asana task: https://app.asana.com/0/1199914526147392/1201177736494461)

### CODE REVIEW

File: packages/snjs/lib/protocol/collection/notes_display_criteria.ts [hunk 1]
```diff
@@ -90,7 +90,15 @@ export class NotesDisplayCriteria {
       }
     }
     if (this.searchQuery) {
-      filters.push((note) => noteMatchesQuery(note, this.searchQuery!));
+      filters.push((note) => {
+        const noteIsAMatch = noteMatchesQuery(note, this.searchQuery!);
+
+        const noteTags = collection.elementsReferencingElement(note)
+          .filter((item) => item.content_type === SNTag) as SNTag[];
```
Comment (johnny243) [inline packages/snjs/lib/protocol/collection/notes_display_criteria.ts:97]: I filtered through the result of `elementsReferencingElement` to make sure all items are of type `SNTag`.

Can a note have references to other items that are not `SNTag`? If so, [this line](https://github.com/standardnotes/snjs/blob/210ba3a0879cb483865469fbbcb3276fad6fa44a/packages/snjs/lib/protocol/collection/notes_display_criteria.ts#L80) should have a safe check.

File: packages/snjs/lib/protocol/collection/notes_display_criteria.ts [hunk 2]
```diff
@@ -90,7 +90,15 @@ export class NotesDisplayCriteria {
       }
     }
     if (this.searchQuery) {
-      filters.push((note) => noteMatchesQuery(note, this.searchQuery!));
+      filters.push((note) => {
+        const noteIsAMatch = noteMatchesQuery(note, this.searchQuery!);
+
+        const noteTags = collection.elementsReferencingElement(note)
+          .filter((item) => item.content_type === ContentType.Tag) as SNTag[];
```
Comment (johnny243) [inline packages/snjs/lib/protocol/collection/notes_display_criteria.ts:97]: I filtered through the result of `elementsReferencingElement` to make sure all items are of type `SNTag`.

Can a note have references to other items that are not `SNTag`? If so, [this line](https://github.com/standardnotes/snjs/blob/210ba3a0879cb483865469fbbcb3276fad6fa44a/packages/snjs/lib/protocol/collection/notes_display_criteria.ts#L80) should have a safe check.

File: packages/snjs/lib/protocol/collection/notes_display_criteria.ts [hunk 3]
```diff
@@ -90,7 +90,15 @@ export class NotesDisplayCriteria {
...
+        const someTagsMatches = noteTags.some((tag) => tagMatchesQuery(tag, this.searchQuery!));
```
Comment (mobitar) [inline packages/snjs/lib/protocol/collection/notes_display_criteria.ts:98]: This should just be moved into `noteMatchesQuery`

File: packages/snjs/mocha/lib/factory.js [hunk 4]
```diff
@@ -380,22 +380,22 @@ export function createNoteParams({ title, text, dirty = true } = {}) {
...
       references: [],
     },
   };
   return params;
 }
 
-export function createRelatedNoteTagPairPayload({ dirty = true } = {}) {
-  const noteParams = createNoteParams({ dirty });
-  const tagParams = createTagParams({ dirty });
+export function createRelatedNoteTagPairPayload({ noteTitle, noteText, tagTitle, dirty = true } = {}) {
```
Comment (mobitar) [inline packages/snjs/mocha/lib/factory.js:396]: Can you review this function/test case? https://github.com/standardnotes/snjs/blob/master/packages/snjs/mocha/payload_encryption.test.js#L106

File: packages/snjs/lib/protocol/collection/notes_display_criteria.ts [hunk 5]
```diff
@@ -155,6 +163,16 @@ export function noteMatchesQuery(
...
 }
 
+export function tagMatchesQuery(
+  tag: SNTag,
+  searchQuery: SearchQuery
+): boolean {
+  if (!searchQuery.query) {
+    return false;
+  }
+  return tag.title.includes(searchQuery.query);
```
Comment (mobitar) [inline packages/snjs/lib/protocol/collection/notes_display_criteria.ts:173]: We'll want case insensitivity here. As well as a test case for it.

File: packages/snjs/lib/protocol/collection/notes_display_criteria.ts [hunk 6]
```diff
@@ -190,6 +201,23 @@ function matchTypeForStringQuery(note: SNNote, searchString: string): Match {
   return (matchesTitle ? Match.Title : 0) + (matchesBody ? Match.Text : 0);
 }
 
+function matchTypeForTagAndStringQuery(tag: SNTag, searchString: string): Match {
+  if (searchString.length === 0) {
+    return Match.Title;
```
Comment (mobitar) [inline packages/snjs/lib/protocol/collection/notes_display_criteria.ts:206]: Should this be Match.None?

**Review by mobitar — APPROVED** [review]

**Pull request merged**